add more columns to arcstat

tkittich opened the pull request:
### Motivation and Context
add more columns to arcstat. Everything from ARC structural breakdown, ARC types breakdown, ARC states breakdown in arc_summary should be in here. Also, I've modified to display decimal places if there's enough space. This will be helpful when interval is large. The values when divided by large interval would become very small. The old behavior would round down small numbers to 0.

### Description
```
   l2wbytes : Bytes written per second to the L2ARC
      cmpsz : Compressed size
     cmpsz% : Compressed size percentage
      ovhsz : Overhead size
     ovhsz% : Overhead size percentage
      bonsz : Bonus size
     bonsz% : Bonus size percentage
      dnosz : Dnode size
     dnosz% : Dnode size percentage
      dbusz : Dbuf size
     dbusz% : Dbuf size percentage
      hdrsz : Header size
     hdrsz% : Header size percentage
      l2hsz : L2 header size
     l2hsz% : L2 header size percentage
      abdsz : ABD chunk waste size
     abdsz% : ABD chunk waste size percentage
     datatg : ARC data target
    datatg% : ARC data target percentage
     datasz : ARC data size
    datasz% : ARC data size percentage
     metatg : ARC metadata target
    metatg% : ARC metadata target percentage
     metasz : ARC metadata size
    metasz% : ARC metadata size percentage
      anosz : Anonymous size
     anosz% : Anonymous size percentage
      anoda : Anonymous data size
     anoda% : Anonymous data size percentage
      anome : Anonymous metadata size
     anome% : Anonymous metadata size percentage
      anoed : Anonymous evictable data size
     anoed% : Anonymous evictable data size percentage
      anoem : Anonymous evictable metadata size
     anoem% : Anonymous evictable metadata size percentage
      mfutg : MFU target
     mfutg% : MFU target percentage
      mfudt : MFU data target
     mfudt% : MFU data target percentage
      mfumt : MFU metadata target
     mfumt% : MFU metadata target percentage
      mfusz : MFU size
     mfusz% : MFU size percentage
      mfuda : MFU data size
     mfuda% : MFU data size percentage
      mfume : MFU metadata size
     mfume% : MFU metadata size percentage
      mfued : MFU evictable data size
     mfued% : MFU evictable data size percentage
      mfuem : MFU evictable metadata size
     mfuem% : MFU evictable metadata size percentage
     mfugsz : MFU ghost size
      mfugd : MFU ghost data size
      mfugm : MFU ghost metadata size
      mrutg : MRU target
     mrutg% : MRU target percentage
      mrudt : MRU data target
     mrudt% : MRU data target percentage
      mrumt : MRU metadata target
     mrumt% : MRU metadata target percentage
      mrusz : MRU size
     mrusz% : MRU size percentage
      mruda : MRU data size
     mruda% : MRU data size percentage
      mrume : MRU metadata size
     mrume% : MRU metadata size percentage
      mrued : MRU evictable data size
     mrued% : MRU evictable data size percentage
      mruem : MRU evictable metadata size
     mruem% : MRU evictable metadata size percentage
     mrugsz : MRU ghost size
      mrugd : MRU ghost data size
      mrugm : MRU ghost metadata size
      uncsz : Uncached size
     uncsz% : Uncached size percentage
      uncda : Uncached data size
     uncda% : Uncached data size percentage
      uncme : Uncached metadata size
     uncme% : Uncached metadata size percentage
      unced : Uncached evictable data size
     unced% : Uncached evictable data size percentage
      uncem : Uncached evictable metadata size
     uncem% : Uncached evictable metadata size percentage
```

### How Has This Been Tested?
Tested locally. 

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

amotin reviewed (commented):
I have no objections for this, but I was thinking that we should add much more of the ARC parameters there.  At very least for completeness there should be also numbers for anon and uncached states.  And then it would be good to see data/metadata distribution, which ARC collects for each state.  And then may be evictable memory stats, etc.

Please see the current `arc_summary` output as an example of what can be reported.  I think `arcstat` should be able to report all the same numbers to allow their analysis in dynamics.

tkittich commented:
I've added more columns as suggested. ^^

```
      anosz : Anonymous size
      mfusz : MFU size
      mrusz : MRU size
      uncsz : Uncached size
    anodata : Anonymous data size
    mfudata : MFU data size
    mrudata : MRU data size
    uncdata : Uncached data size
    anometa : Anonymous metadata size
    mfumeta : MFU metadata size
    mrumeta : MRU metadata size
    uncmeta : Uncached metadata size
   anoedata : Anonymous evictable data size
   mfuedata : MFU evictable data size
   mruedata : MRU evictable data size
   uncedata : Uncached evictable data size
   anoemeta : Anonymous evictable metadata size
   mfuemeta : MFU evictable metadata size
   mruemeta : MRU evictable metadata size
   uncemeta : Uncached evictable metadata size
```

amotin reviewed (approved):
Thank you. I can think of more counters, like for ghost states (even though they are quite different and may be confusing) or total data/metadata size, structural breakdown section of arc_summary (dnodes, etc), but this is definitely a good addition.

tkittich commented:
I've add more columns. Everything from ARC structural breakdown, ARC types breakdown, ARC states breakdown in arc_summary should be in here.

```
   l2wbytes : Bytes written per second to the L2ARC
       cpsz : Compressed size
      cpsz% : Compressed size percentage
       ohsz : Overhead size
      ohsz% : Overhead size percentage
       bnsz : Bonus size
      bnsz% : Bonus size percentage
       dnsz : Dnode size
      dnsz% : Dnode size percentage
       dbsz : Dbuf size
      dbsz% : Dbuf size percentage
      hdrsz : Header size
     hdrsz% : Header size percentage
      l2hsz : L2 header size
     l2hsz% : L2 header size percentage
      adbsz : ABD chunk waste size
     adbsz% : ABD chunk waste size percentage
     datasz : ARC data size
    datasz% : ARC data size percentage
     metasz : ARC metadata size
    metasz% : ARC metadata size percentage
      anosz : Anonymous size
     anosz% : Anonymous size percentage
    anodata : Anonymous data size
   anodata% : Anonymous data size percentage
    anometa : Anonymous metadata size
   anometa% : Anonymous metadata size percentage
   anoedata : Anonymous evictable data size
  anoedata% : Anonymous evictable data size percentage
   anoemeta : Anonymous evictable metadata size
  anoemeta% : Anonymous evictable metadata size percentage
      mfusz : MFU size
     mfusz% : MFU size percentage
    mfudata : MFU data size
   mfudata% : MFU data size percentage
    mfumeta : MFU metadata size
   mfumeta% : MFU metadata size percentage
   mfuedata : MFU evictable data size
  mfuedata% : MFU evictable data size percentage
   mfuemeta : MFU evictable metadata size
  mfuemeta% : MFU evictable metadata size percentage
     mfugsz : MFU ghost size
   mfugdata : MFU ghost data size
   mfugmeta : MFU ghost metadata size
      mrusz : MRU size
     mrusz% : MRU size percentage
    mrudata : MRU data size
   mrudata% : MRU data size percentage
    mrumeta : MRU metadata size
   mrumeta% : MRU metadata size percentage
   mruedata : MRU evictable data size
  mruedata% : MRU evictable data size percentage
   mruemeta : MRU evictable metadata size
  mruemeta% : MRU evictable metadata size percentage
     mrugsz : MRU ghost size
   mrugdata : MRU ghost data size
   mrugmeta : MRU ghost metadata size
      uncsz : Uncached size
     uncsz% : Uncached size percentage
    uncdata : Uncached data size
   uncdata% : Uncached data size percentage
    uncmeta : Uncached metadata size
   uncmeta% : Uncached metadata size percentage
   uncedata : Uncached evictable data size
  uncedata% : Uncached evictable data size percentage
   uncemeta : Uncached evictable metadata size
  uncemeta% : Uncached evictable metadata size percentage
```

amotin commented on cmd/arcstat.in:
```diff
@@ -169,6 +170,74 @@ cols = {
+                    continue
+                colname = field + stat
+                coldesc = fieldval[1] + " " + statval[1]
+                cols[colname] = [len(colname), 1024, coldesc]
```
The column width should account not only name length, but also value length, which for 1024-base may be 5, while this code I suppose will produce 4 for `cpsz`, etc.

amotin reviewed (approved):
That is quite a list and quite a logic. ;)  Just one cosmetic comment:

amotin commented:
And please squash the commits.

tkittich commented:
> And please squash the commits.

Thank you for the review. I've added target sizes and changed column names to fit 5 chars. ^^ However, it seems I [don't have authorization](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github#squashing-your-merge-commits) to squash the commits.

amotin commented on cmd/arcstat.in:
```diff
@@ -506,7 +600,8 @@ def calculate():
     global l2exist
-
+    global fieldstats
+    
```
Was here some unneeded spaces before or are there now?  I.e. is this a fix or a new typo?

amotin reviewed (approved):
Thanks.

tkittich commented:
I've added target size of ARC data, ARC metadata, MFU, MRU. ^^

amotin reviewed (approved):
Thanks.  I am slightly worrying that those grouped numbers could be confusing, showing that target is not reached, while one of their components actually is, but more is not less.

allanjude commented:
Anon_size is definitely the one I've wanted many times

tkittich commented:
I've shorten new column names. Also, I've modified to display decimal places when number is < 10. This will be helpful when interval is large. The values when divided by large interval would become very small. The old behavior would round down small numbers to 0.
```
      cmpsz : Compressed size
     cmpsz% : Compressed size percentage
      ovhsz : Overhead size
     ovhsz% : Overhead size percentage
      bonsz : Bonus size
     bonsz% : Bonus size percentage
      dnosz : Dnode size
     dnosz% : Dnode size percentage
      dbusz : Dbuf size
     dbusz% : Dbuf size percentage
      hdrsz : Header size
     hdrsz% : Header size percentage
      l2hsz : L2 header size
     l2hsz% : L2 header size percentage
      abdsz : ABD chunk waste size
     abdsz% : ABD chunk waste size percentage
     datatg : ARC data target
    datatg% : ARC data target percentage
     datasz : ARC data size
    datasz% : ARC data size percentage
     metatg : ARC metadata target
    metatg% : ARC metadata target percentage
     metasz : ARC metadata size
    metasz% : ARC metadata size percentage
      anosz : Anonymous size
     anosz% : Anonymous size percentage
      anoda : Anonymous data size
     anoda% : Anonymous data size percentage
      anome : Anonymous metadata size
     anome% : Anonymous metadata size percentage
      anoed : Anonymous evictable data size
     anoed% : Anonymous evictable data size percentage
      anoem : Anonymous evictable metadata size
     anoem% : Anonymous evictable metadata size percentage
      mfutg : MFU target
     mfutg% : MFU target percentage
      mfudt : MFU data target
     mfudt% : MFU data target percentage
      mfumt : MFU metadata target
     mfumt% : MFU metadata target percentage
      mfusz : MFU size
     mfusz% : MFU size percentage
      mfuda : MFU data size
     mfuda% : MFU data size percentage
      mfume : MFU metadata size
     mfume% : MFU metadata size percentage
      mfued : MFU evictable data size
     mfued% : MFU evictable data size percentage
      mfuem : MFU evictable metadata size
     mfuem% : MFU evictable metadata size percentage
     mfugsz : MFU ghost size
      mfugd : MFU ghost data size
      mfugm : MFU ghost metadata size
      mrutg : MRU target
     mrutg% : MRU target percentage
      mrudt : MRU data target
     mrudt% : MRU data target percentage
      mrumt : MRU metadata target
     mrumt% : MRU metadata target percentage
      mrusz : MRU size
     mrusz% : MRU size percentage
      mruda : MRU data size
     mruda% : MRU data size percentage
      mrume : MRU metadata size
     mrume% : MRU metadata size percentage
      mrued : MRU evictable data size
     mrued% : MRU evictable data size percentage
      mruem : MRU evictable metadata size
     mruem% : MRU evictable metadata size percentage
     mrugsz : MRU ghost size
      mrugd : MRU ghost data size
      mrugm : MRU ghost metadata size
      uncsz : Uncached size
     uncsz% : Uncached size percentage
      uncda : Uncached data size
     uncda% : Uncached data size percentage
      uncme : Uncached metadata size
     uncme% : Uncached metadata size percentage
      unced : Uncached evictable data size
     unced% : Uncached evictable data size percentage
      uncem : Uncached evictable metadata size
     uncem% : Uncached evictable metadata size percentage
```

amotin commented:
Fractional bytes or operations may weird is they are not percents or per-second. ;)

tkittich commented:
> Fractional bytes or operations may weird is they are not percents or per-second. ;)

The latest commit has a fix that will only show decimal places when (interval is 1 and value is < 1) or (interval is not 1 and value is < 10) . ^^
